core/src/main/java/io/parsingdata/metal/token/Until.java

```diff
@@ -0,0 +1,131 @@
+        this.stepSize = stepSize == null ? DEFAULT_STEP : stepSize;
+        this.maxSize = maxSize == null ? DEFAULT_MAX : maxSize;
+        this.terminator = checkNotNull(terminator, "terminator");
+    }
+
+    @Override
+    protected Optional<Environment> parseImpl(final String scope, final Environment environment, final Encoding encoding) throws IOException {
+        return handleInterval(scope, environment, initialSize.eval(environment.order, encoding), stepSize.eval(environment.order, encoding), maxSize.eval(environment.order, encoding), encoding).computeResult();
+    }
+
+    private Trampoline<Optional<Environment>> handleInterval(final String scope, final Environment environment, final ImmutableList<Optional<Value>> initialSizes, final ImmutableList<Optional<Value>> stepSizes, final ImmutableList<Optional<Value>> maxSizes, final Encoding encoding) throws IOException {
+        if (checkNotValidList(initialSizes) || checkNotValidList(stepSizes) || checkNotValidList(maxSizes)) { return complete(Util::failure); }
+        return iterate(scope, environment, getInt(initialSizes), getInt(stepSizes), getInt(maxSizes), encoding).computeResult()
+            .map(nextEnvironment -> complete(() -> success(nextEnvironment)))
+            .orElseGet(() -> intermediate(() -> handleInterval(scope, environment, initialSizes.tail, stepSizes.tail, maxSizes.tail, encoding)));
+    }
+
+    private Trampoline<Optional<Environment>> iterate(final String scope, final Environment environment, final int currentSize, final int stepSize, final int maxSize, final Encoding encoding) throws IOException {
+        if (stepSize == 0 || stepSize > 0 && currentSize > maxSize || stepSize < 0 && currentSize < maxSize) { return complete(Util::failure); }
+        final Slice slice = environment.slice(currentSize);
+        if (slice.size != currentSize) { return complete(Util::failure); }
+        return terminator.parse(scope, currentSize == 0 ? environment : environment.add(new ParseValue(name, this, slice, encoding)).seek(environment.offset + currentSize), encoding)
+            .map(nextEnvironment -> complete(() -> success(nextEnvironment)))
+            .orElseGet(() -> intermediate(() -> iterate(scope, environment, currentSize + stepSize, stepSize, maxSize, encoding)));
+    }
+
+    private boolean checkNotValidList(ImmutableList<Optional<Value>> list) {
+        return list.isEmpty() || !list.head.isPresent();
+    }
+
+    private int getInt(ImmutableList<Optional<Value>> list) {
+        return list.head.get().asNumeric().intValue();
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + "(" + makeNameFragment() + initialSize + "," + stepSize + "," + maxSize + "," + terminator + ")";
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        return super.equals(obj)
+            && Objects.equals(initialSize, ((Until)obj).initialSize)
```

Comment:
Is this safe? What if obj is an instance of an other subclass but equal to super? Then these casts will fail?